group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/AbstractUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,260 @@
+    ) {
+        List<TopicIdPartition> allTopicIdPartitions = new ArrayList<>();
+        // Sorted so that partitions from each topic can be distributed amongst its subscribers equally.
+        allTopicIds.stream().sorted().forEach(topic ->
+            IntStream.range(0, subscribedTopicDescriber.numPartitions(topic))
+                .forEach(i -> allTopicIdPartitions.add(new TopicIdPartition(topic, i)))
+        );
+
+        return allTopicIdPartitions;
+    }
+
+    /**
+     * Represents the rack information of members and partitions along with utility methods
+     * to facilitate rack-aware assignment strategies for a given consumer group.
+     */
+    protected static class RackInfo {
+        /**
+         * Map of every member to its rack.
+         */
+        protected final Map<String, String> memberRacks;
+
+        /**
+         * Map of every partition to a list of its racks.
+         */
+        protected final Map<TopicIdPartition, Set<String>> partitionRacks;
+
+        /**
+         * List of members with the same rack as the partition.
+         */
+        protected final Map<TopicIdPartition, List<String>> membersWithSameRackAsPartition;
+
+        /**
+         * Indicates if a rack aware assignment can be done.
+         * True if racks are defined for both members and partitions and there is an intersection between the sets.
+         */
+        protected final boolean useRackStrategy;
+
+        /**
+         * Constructs rack information based on the assignment specification and subscribed topics.
+         *
+         * @param assignmentSpec                The current assignment specification.
+         * @param subscribedTopicDescriber      Topic and partition metadata of the subscribed topics.
+         * @param topicIds                      List of topic Ids.
+         */
+        public RackInfo(
+            AssignmentSpec assignmentSpec,
+            SubscribedTopicDescriber subscribedTopicDescriber,
+            Set<Uuid> topicIds
+        ) {
+            Map<String, List<String>> membersByRack = new HashMap<>();
+            assignmentSpec.members().forEach((memberId, assignmentMemberSpec) ->
+                assignmentMemberSpec.rackId().filter(r -> !r.isEmpty()).ifPresent(
+                    rackId -> membersByRack.computeIfAbsent(rackId, __ -> new ArrayList<>()).add(memberId)
+                )
+            );
+
+            Set<String> allPartitionRacks;
+            Map<TopicIdPartition, Set<String>> racksPerPartition;
+            List<TopicIdPartition> topicIdPartitions = allTopicIdPartitions(topicIds, subscribedTopicDescriber);
+
+            if (membersByRack.isEmpty()) {
+                allPartitionRacks = Collections.emptySet();
+                racksPerPartition = Collections.emptyMap();
+            } else {
+                racksPerPartition = new HashMap<>();
+                allPartitionRacks = new HashSet<>();
+                topicIdPartitions.forEach(tp -> {
```

Comment:
It is actually a bit wasteful to generate a List to just iterate once on it, no?

Reply:
yes agreed, I changed it now, thanks!